Update of the DPL I/O API

matthiasrichter opened the pull request:
- added support for ROOT serialization of non-TObject objects
- automatically deduce serialization method for messageable and ROOT-serializable objects and handle ambiguity for those which are both
- adding type wrapper to force ROOT-serialization method

alibuild commented:
Error while checking build/O2/o2 for 453535bc93efe102f413f581d975efe6ea658228:
```
sw/BUILD/O2-latest/log
#undef  _BACKWARD_BACKWARD_WARNING_H

In file included from input_line_10:1:
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:25:19: error: no member named 'Cluster' in namespace 'o2::ITSMFT'
using o2::ITSMFT::Cluster;
      ~~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/83735733f4a25f7adb6a3653a337a8caed8f914c/O2/lib/libITSMFTSimulation.rootmap declaring '
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:32:21: error: no member named 'Hit' in namespace 'o2::ITSMFT'
  using o2::ITSMFT::Hit;
        ~~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/83735733f4a25f7adb6a3653a337a8caed8f914c/O2/lib/libMFTBase.rootmap declaring '
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:47:24: error: no member named 'GeometryTGeo' in namespace 'o2::MFT'
  auto gman = o2::MFT::GeometryTGeo::Instance();
              ~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:48:36: error: no member named 'bit2Mask' in namespace 'o2::utils'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                        ~~~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:48:49: error: no member named 'TransformType' in namespace 'o2'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                                            ~~~~^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:48:73: error: no member named 'TransformType' in namespace 'o2'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                                                                    ~~~~^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:49:49: error: no member named 'TransformType' in namespace 'o2'
                                            o2::TransformType::L2G)); // request cached transforms
                                            ~~~~^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:55:15: error: use of undeclared identifier 'Hit'
  std::vector<Hit>* hitArray = nullptr;
              ^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:55:21: error: C++ requires a type specifier for all declarations
  std::vector<Hit>* hitArray = nullptr;
                    ^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:62:15: error: use of undeclared identifier 'Cluster'
  std::vector<Cluster>* clusArray = nullptr;
              ^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:62:25: error: C++ requires a type specifier for all declarations
  std::vector<Cluster>* clusArray = nullptr;
                        ^
G__MathUtilsDict dictionary payload:1363:41: error: cannot initialize a member subobject of type 'std::vector<double> *' with an rvalue of type 'std::vector<Double_t> *' (aka 'vector<double> *')
  std::vector<double>* mIntegralCache = &fIntegral;
                                        ^~~~~~~~~~
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/867/453535bc93efe102f413f581d975efe6ea658228/fullLog.txt).

alibuild commented:
Error while checking build/o2checkcode/o2 for 453535bc93efe102f413f581d975efe6ea658228:
```
sw/BUILD/O2-latest/log
47/70 Test #49: test_Framework_test_TypeTraits .....................   Passed    0.65 sec
      Start 47: test_Framework_test_SuppressionGenerator
      Start 57: test_Algorithm_parser
48/70 Test #63: test_DataFlow_test_SubframeUtils01 .................   Passed    0.87 sec
      Start 51: test_Framework_test_WorkflowHelpers
49/70 Test #36: test_Framework_test_DataProcessorSpec ..............   Passed    0.76 sec
      Start  7: test_Headers_testTimeStamp
50/70 Test #47: test_Framework_test_SuppressionGenerator ...........   Passed    0.63 sec
51/70 Test #57: test_Algorithm_parser ..............................   Passed    0.63 sec
      Start 16: test_DataFormatsTPC_ClusterNative
      Start 58: test_Algorithm_tableview
52/70 Test #51: test_Framework_test_WorkflowHelpers ................   Passed    0.82 sec
      Start 60: test_aliceHLTwrapper_testMessageFormat
53/70 Test  #7: test_Headers_testTimeStamp .........................   Passed    0.63 sec
      Start  8: test_Headers_test_HeartbeatFrame
54/70 Test #16: test_DataFormatsTPC_ClusterNative ..................   Passed    0.65 sec
55/70 Test #58: test_Algorithm_tableview ...........................   Passed    0.75 sec
      Start 56: test_Algorithm_headerstack
      Start  2: test_MathUtils_testCartesian3D
56/70 Test #60: test_aliceHLTwrapper_testMessageFormat .............   Passed    0.86 sec
      Start 59: test_Algorithm_pageparser
57/70 Test  #8: test_Headers_test_HeartbeatFrame ...................   Passed    0.65 sec
      Start 15: test_DetectorsCommonDataFormats_testDetID
58/70 Test  #2: test_MathUtils_testCartesian3D .....................   Passed    0.62 sec
59/70 Test #56: test_Algorithm_headerstack .........................   Passed    0.72 sec
      Start  9: test_Headers_test_RAWDataHeader
      Start 55: test_Algorithm_o2formatparser
60/70 Test #59: test_Algorithm_pageparser ..........................   Passed    0.83 sec
      Start  6: test_Headers_testDataHeader
61/70 Test #15: test_DetectorsCommonDataFormats_testDetID ..........   Passed    0.64 sec
      Start 64: test_DataFlow_test_PayloadMerger01
62/70 Test #55: test_Algorithm_o2formatparser ......................   Passed    0.64 sec
63/70 Test  #9: test_Headers_test_RAWDataHeader ....................   Passed    0.74 sec
      Start 68: test_DataCompression_test_HuffmanCodec
      Start 17: test_CommonDataFormat_testTimeStamp
64/70 Test #17: test_CommonDataFormat_testTimeStamp ................   Passed    0.03 sec
65/70 Test  #6: test_Headers_testDataHeader ........................   Passed    0.77 sec
      Start 65: test_DataCompression_test_dc_primitives
      Start 69: test_DataCompression_test_DataDeflater
66/70 Test #64: test_DataFlow_test_PayloadMerger01 .................   Passed    0.60 sec
67/70 Test #65: test_DataCompression_test_dc_primitives ............   Passed    0.16 sec
68/70 Test #69: test_DataCompression_test_DataDeflater .............   Passed    0.13 sec
      Start 54: test_Framework_test_DataAllocator
69/70 Test #68: test_DataCompression_test_HuffmanCodec .............   Passed    0.61 sec
70/70 Test #54: test_Framework_test_DataAllocator ..................   Passed    0.64 sec

100% tests passed, 0 tests failed out of 70

Total Test time (real) =  21.95 sec
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/867/453535bc93efe102f413f581d975efe6ea658228/fullLog.txt).

alibuild commented:
Error while checking build/o2/macos for 453535bc93efe102f413f581d975efe6ea658228:
```
sw/BUILD/O2-latest/log
Scanning dependencies of target test_Framework_test_LogParsingHelpers
Scanning dependencies of target test_Framework_test_SingleDataSource
In file included from /build/ci_checks/o2_checker_macos_o2_ci_0/sw/SOURCES/O2/867/0/Framework/Core/test/test_TypeTraits.cxx:15:
/build/ci_checks/o2_checker_macos_o2_ci_0/sw/SOURCES/O2/867/0/Framework/Core/include/Framework/TypeTraits.h:95:34: error: call to 'test' is ambiguous
  enum { value = sizeof(decltype(test<T>(nullptr))) == sizeof(Yes) };
                                 ^~~~~~~
/build/ci_checks/o2_checker_macos_o2_ci_0/sw/SOURCES/O2/867/0/Framework/Core/test/test_TypeTraits.cxx:93:23: note: in instantiation of template class 'o2::framework::has_root_dictionary<o2::test::TriviallyCopyable, void>' requested here
--
--
/build/ci_checks/o2_checker_macos_o2_ci_0/sw/SOURCES/O2/867/0/Framework/Core/include/Framework/TypeTraits.h:92:13: note: candidate function [with U = o2::test::TriviallyCopyable]
  static No test(U*);
            ^
/build/ci_checks/o2_checker_macos_o2_ci_0/sw/SOURCES/O2/867/0/Framework/Core/include/Framework/TypeTraits.h:95:34: error: call to 'test' is ambiguous
  enum { value = sizeof(decltype(test<T>(nullptr))) == sizeof(Yes) };
                                 ^~~~~~~
/build/ci_checks/o2_checker_macos_o2_ci_0/sw/SOURCES/O2/867/0/Framework/Core/test/test_TypeTraits.cxx:94:23: note: in instantiation of template class 'o2::framework::has_root_dictionary<o2::test::Polymorphic, void>' requested here
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/867/453535bc93efe102f413f581d975efe6ea658228/fullLog.txt).

alibuild commented:
Error while checking build/O2/o2-dev-fairroot for 453535bc93efe102f413f581d975efe6ea658228:
```

```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/867/453535bc93efe102f413f581d975efe6ea658228/fullLog.txt).

alibuild commented:
Error while checking build/O2/o2-dev-fairroot for 453535bc93efe102f413f581d975efe6ea658228:
```
sw/BUILD/O2-latest/log
#undef  _BACKWARD_BACKWARD_WARNING_H

In file included from input_line_10:1:
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:24:19: error: no member named 'Cluster' in namespace 'o2::ITSMFT'
using o2::ITSMFT::Cluster;
      ~~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/d5673ec7b83d225351df0ca7ef11a16203c6c0c1/O2/lib/libITSMFTSimulation.rootmap declaring '
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:31:21: error: no member named 'Hit' in namespace 'o2::ITSMFT'
  using o2::ITSMFT::Hit;
        ~~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/d5673ec7b83d225351df0ca7ef11a16203c6c0c1/O2/lib/libMFTBase.rootmap declaring '
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:46:24: error: no member named 'GeometryTGeo' in namespace 'o2::MFT'
  auto gman = o2::MFT::GeometryTGeo::Instance();
              ~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:47:36: error: no member named 'bit2Mask' in namespace 'o2::utils'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                        ~~~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:47:49: error: no member named 'TransformType' in namespace 'o2'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                                            ~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:47:73: error: no member named 'TransformType' in namespace 'o2'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                                                                    ~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:48:49: error: no member named 'TransformType' in namespace 'o2'
                                            o2::TransformType::L2G)); // request cached transforms
                                            ~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:54:15: error: use of undeclared identifier 'Hit'
  std::vector<Hit>* hitArray = nullptr;
              ^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:54:21: error: C++ requires a type specifier for all declarations
  std::vector<Hit>* hitArray = nullptr;
                    ^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:61:15: error: use of undeclared identifier 'Cluster'
  std::vector<Cluster>* clusArray = nullptr;
              ^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:61:25: error: C++ requires a type specifier for all declarations
  std::vector<Cluster>* clusArray = nullptr;
                        ^
Error in <TInterpreter::AutoParse>: Error parsing payload code for class o2::Rotation2D with content:
--

```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/867/453535bc93efe102f413f581d975efe6ea658228/fullLog.txt).

ktf commented on Framework/Core/CMakeLists.txt:
```diff
@@ -51,11 +51,18 @@ set(SRCS
     src/runDataProcessing.cxx
     src/TMessageSerializer.cxx
     ${GUI_SOURCES}
+    test/Classes.cxx
```
How about we use a less specific name?

matthiasrichter replied:
like TestClasses?

ktf replied:
+1

ktf commented on Framework/Core/include/Framework/InputRecord.h:
```diff
@@ -29,11 +29,74 @@
+{
+ public:
+  enum class OwnershipProperty : short {
+    Unknown = -1,
```
I think our enums should be all capital. Also I would use `enum struct` rather than `enum class`.

matthiasrichter replied:
ok

matthiasrichter replied:
will change to `enum struct`, as for the names the coding conventions say camel case convention, so they should be fine

ktf commented on Framework/Core/include/Framework/InputRecord.h:
```diff
@@ -29,11 +29,74 @@
+/// InputRecord and DPL processing APIs. The Deleter can be extended
+/// to support a callback to call a resource management outside.
+template <typename T>
+class Deleter : public default_delete<T>
```
I would use a more explicative name than simply `Deleter`. Notice also that this effectively means we will be using an extra int per pointer. Is that reasonable ok?

matthiasrichter replied:
Agreed, how about moving `Deleter` into `InputRecord`, than it becomes `InputRecord::Deleter` which is probably explicative enough?

Regarding the increased size for a pointer I don't think its and issue as this is per incoming data blocks, definitely less than O(1k).

ktf replied:
Sounds good.

ktf commented on Framework/Core/include/Framework/SerializationMethods.h:
```diff
@@ -0,0 +1,66 @@
+/// Currently the main purpose of the base class is to make the converters
+/// polymorphic and distinguishable from the messageable types in
+/// automatic substitutions
+class SerializationMethod
```
I would make it a struct and avoid any virtual method. Why do you need them? Actually, why do you need this at all? Given they are all templated, the various serialization methods do not need any polymorphism.

matthiasrichter replied:
In order to avoid ambiguous substitutions I had to make this converter non- messageable. The  comment just in front of the definition tries to explain this.

We would need to add is-not-ROOTSerialized like statements in all the specializations and possibly the same for other converters coming in the future. The simple solution is to make the converters non-messageable, but maybe there are better ways.

ktf replied:
Ah, ok, I see. How about we check for a `::non_messageable` type when checking if `is_messageable`. Kinda like when you do `is_container` by checking the types which are known to be required for a `std::container`.

ktf replied:
Actual example of what I mean:

```C++
#include <iostream>
#include <type_traits>

struct NonMessageable {};

struct is_non_messageable {
  template <typename T>
  static typename std::enable_if<std::is_same<typename T::non_messageable, NonMessageable>::value, bool>::type docheck(int) {
    return true;
  }
  template <typename T>
  static bool docheck(long) {
    return false;
  }
};

struct Foo {
  using non_messageable = NonMessageable;
};

int main(int argc, char**argv) {
  std::cerr << is_non_messageable::docheck<Foo>(0) << std::endl;
  std::cerr << is_non_messageable::docheck<int>(0) << std::endl;
}
```

can probably probably be included, but should give an idea of what I meant.

alibuild commented:
Error while checking build/O2/o2-dev-fairroot for 6e435a19a5627be76017e4528ac53bfda49f8073:
```
sw/BUILD/O2-latest/log
#undef  _BACKWARD_BACKWARD_WARNING_H

In file included from input_line_10:1:
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:24:19: error: no member named 'Cluster' in namespace 'o2::ITSMFT'
using o2::ITSMFT::Cluster;
      ~~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/7e2770c80a55ac267e8e5007c180a91fef4be384/O2/lib/libITSMFTSimulation.rootmap declaring '
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:31:21: error: no member named 'Hit' in namespace 'o2::ITSMFT'
  using o2::ITSMFT::Hit;
        ~~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/7e2770c80a55ac267e8e5007c180a91fef4be384/O2/lib/libMFTBase.rootmap declaring '
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:46:24: error: no member named 'GeometryTGeo' in namespace 'o2::MFT'
  auto gman = o2::MFT::GeometryTGeo::Instance();
              ~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:47:36: error: no member named 'bit2Mask' in namespace 'o2::utils'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                        ~~~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:47:49: error: no member named 'TransformType' in namespace 'o2'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                                            ~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:47:73: error: no member named 'TransformType' in namespace 'o2'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                                                                    ~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:48:49: error: no member named 'TransformType' in namespace 'o2'
                                            o2::TransformType::L2G)); // request cached transforms
                                            ~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:54:15: error: use of undeclared identifier 'Hit'
  std::vector<Hit>* hitArray = nullptr;
              ^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:54:21: error: C++ requires a type specifier for all declarations
  std::vector<Hit>* hitArray = nullptr;
                    ^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:61:15: error: use of undeclared identifier 'Cluster'
  std::vector<Cluster>* clusArray = nullptr;
              ^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:61:25: error: C++ requires a type specifier for all declarations
  std::vector<Cluster>* clusArray = nullptr;
                        ^
Error in <TInterpreter::AutoParse>: Error parsing payload code for class o2::Rotation2D with content:
--

```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/867/6e435a19a5627be76017e4528ac53bfda49f8073/fullLog.txt).

alibuild commented:
Error while checking build/o2checkcode/o2 for 6e435a19a5627be76017e4528ac53bfda49f8073:
```

```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/867/6e435a19a5627be76017e4528ac53bfda49f8073/fullLog.txt).

matthiasrichter commented:
@ktf @dberzano could you trigger the macos ci for this PR, thanks.

alibuild commented:
Error while checking build/O2/o2-dev-fairroot for b87a58cac4c41090b57ffab2edb6f2d7be9c5cf9:
```
sw/BUILD/O2-latest/log
#undef  _BACKWARD_BACKWARD_WARNING_H

In file included from input_line_10:1:
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:25:19: error: no member named 'Cluster' in namespace 'o2::ITSMFT'
using o2::ITSMFT::Cluster;
      ~~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/7e2770c80a55ac267e8e5007c180a91fef4be384/O2/lib/libITSMFTSimulation.rootmap declaring '
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:32:21: error: no member named 'Hit' in namespace 'o2::ITSMFT'
  using o2::ITSMFT::Hit;
        ~~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/7e2770c80a55ac267e8e5007c180a91fef4be384/O2/lib/libMFTBase.rootmap declaring '
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:47:24: error: no member named 'GeometryTGeo' in namespace 'o2::MFT'
  auto gman = o2::MFT::GeometryTGeo::Instance();
              ~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:48:36: error: no member named 'bit2Mask' in namespace 'o2::utils'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                        ~~~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:48:49: error: no member named 'TransformType' in namespace 'o2'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                                            ~~~~^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:48:73: error: no member named 'TransformType' in namespace 'o2'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                                                                    ~~~~^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:49:49: error: no member named 'TransformType' in namespace 'o2'
                                            o2::TransformType::L2G)); // request cached transforms
                                            ~~~~^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:55:15: error: use of undeclared identifier 'Hit'
  std::vector<Hit>* hitArray = nullptr;
              ^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:55:21: error: C++ requires a type specifier for all declarations
  std::vector<Hit>* hitArray = nullptr;
                    ^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:62:15: error: use of undeclared identifier 'Cluster'
  std::vector<Cluster>* clusArray = nullptr;
              ^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:62:25: error: C++ requires a type specifier for all declarations
  std::vector<Cluster>* clusArray = nullptr;
                        ^
Error in <TInterpreter::AutoParse>: Error parsing payload code for class o2::Rotation2D with content:
--

```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/867/b87a58cac4c41090b57ffab2edb6f2d7be9c5cf9/fullLog.txt).

alibuild commented:
Error while checking build/O2/o2-dev-fairroot for 76f9b5b5368c152aabb3b21ecfa139f420311529:
```
sw/BUILD/O2-latest/log
#undef  _BACKWARD_BACKWARD_WARNING_H

In file included from input_line_10:1:
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:25:19: error: no member named 'Cluster' in namespace 'o2::ITSMFT'
using o2::ITSMFT::Cluster;
      ~~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/7e2770c80a55ac267e8e5007c180a91fef4be384/O2/lib/libITSMFTSimulation.rootmap declaring '
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:32:21: error: no member named 'Hit' in namespace 'o2::ITSMFT'
  using o2::ITSMFT::Hit;
        ~~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/7e2770c80a55ac267e8e5007c180a91fef4be384/O2/lib/libMFTBase.rootmap declaring '
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:47:24: error: no member named 'GeometryTGeo' in namespace 'o2::MFT'
  auto gman = o2::MFT::GeometryTGeo::Instance();
              ~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:48:36: error: no member named 'bit2Mask' in namespace 'o2::utils'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                        ~~~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:48:49: error: no member named 'TransformType' in namespace 'o2'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                                            ~~~~^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:48:73: error: no member named 'TransformType' in namespace 'o2'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                                                                    ~~~~^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:49:49: error: no member named 'TransformType' in namespace 'o2'
                                            o2::TransformType::L2G)); // request cached transforms
                                            ~~~~^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:55:15: error: use of undeclared identifier 'Hit'
  std::vector<Hit>* hitArray = nullptr;
              ^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:55:21: error: C++ requires a type specifier for all declarations
  std::vector<Hit>* hitArray = nullptr;
                    ^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:62:15: error: use of undeclared identifier 'Cluster'
  std::vector<Cluster>* clusArray = nullptr;
              ^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/MFT/macros/CheckClusters.C:62:25: error: C++ requires a type specifier for all declarations
  std::vector<Cluster>* clusArray = nullptr;
                        ^
Error in <TInterpreter::AutoParse>: Error parsing payload code for class o2::Rotation2D with content:
--

```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/867/76f9b5b5368c152aabb3b21ecfa139f420311529/fullLog.txt).

matthiasrichter commented:
@ktf, I think I have addressed all your comments. For making the `ROOTSerialized` wrapper a non-messageable  type I have introduced some type trait following your idea.
```c++
struct MarkAsNonMessageable {};

template< typename T, typename _ = void >
struct force_non_messageable : public std::false_type {};

template <typename T>
struct force_non_messageable<
  T,
  std::conditional_t<
    false,
    class_member_checker<
      typename T::non_messageable,
      typename std::enable_if<std::is_same<typename T::non_messageable, MarkAsNonMessageable>::value>::type
      >,
    void
    >
  > : public std::true_type {};

```
this follows the same idea as the detection of std containers. `force_non_messageable` should be probably `is_forced_non_messageable`.

I still need to clean up the PR.

ktf commented:
Looks great, thanks. Once you are done I think we can merge this.

alibuild commented:
Error while checking build/O2/o2-dev-fairroot for 8dda4f9c2da84555189d62440af875f25c1e016e:
```
sw/BUILD/O2-latest/log
#undef  _BACKWARD_BACKWARD_WARNING_H

In file included from input_line_10:1:
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:24:19: error: no member named 'Cluster' in namespace 'o2::ITSMFT'
using o2::ITSMFT::Cluster;
      ~~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/7e2770c80a55ac267e8e5007c180a91fef4be384/O2/lib/libITSMFTSimulation.rootmap declaring '
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:31:21: error: no member named 'Hit' in namespace 'o2::ITSMFT'
  using o2::ITSMFT::Hit;
        ~~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/7e2770c80a55ac267e8e5007c180a91fef4be384/O2/lib/libMFTBase.rootmap declaring '
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:46:24: error: no member named 'GeometryTGeo' in namespace 'o2::MFT'
  auto gman = o2::MFT::GeometryTGeo::Instance();
              ~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:47:36: error: no member named 'bit2Mask' in namespace 'o2::utils'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                        ~~~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:47:49: error: no member named 'TransformType' in namespace 'o2'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                                            ~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:47:73: error: no member named 'TransformType' in namespace 'o2'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                                                                    ~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:48:49: error: no member named 'TransformType' in namespace 'o2'
                                            o2::TransformType::L2G)); // request cached transforms
                                            ~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:54:15: error: use of undeclared identifier 'Hit'
  std::vector<Hit>* hitArray = nullptr;
              ^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:54:21: error: C++ requires a type specifier for all declarations
  std::vector<Hit>* hitArray = nullptr;
                    ^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:61:15: error: use of undeclared identifier 'Cluster'
  std::vector<Cluster>* clusArray = nullptr;
              ^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:61:25: error: C++ requires a type specifier for all declarations
  std::vector<Cluster>* clusArray = nullptr;
                        ^
Error in <TInterpreter::AutoParse>: Error parsing payload code for class o2::Rotation2D with content:
--

```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/867/8dda4f9c2da84555189d62440af875f25c1e016e/fullLog.txt).

alibuild commented:
Error while checking build/O2/o2-dev-fairroot for 378e7b83097db8555e3a8f289e609afc6785f259:
```
sw/BUILD/O2-latest/log
#undef  _BACKWARD_BACKWARD_WARNING_H

In file included from input_line_10:1:
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:24:19: error: no member named 'Cluster' in namespace 'o2::ITSMFT'
using o2::ITSMFT::Cluster;
      ~~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/7e2770c80a55ac267e8e5007c180a91fef4be384/O2/lib/libITSMFTSimulation.rootmap declaring '
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:31:21: error: no member named 'Hit' in namespace 'o2::ITSMFT'
  using o2::ITSMFT::Hit;
        ~~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/7e2770c80a55ac267e8e5007c180a91fef4be384/O2/lib/libMFTBase.rootmap declaring '
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:46:24: error: no member named 'GeometryTGeo' in namespace 'o2::MFT'
  auto gman = o2::MFT::GeometryTGeo::Instance();
              ~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:47:36: error: no member named 'bit2Mask' in namespace 'o2::utils'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                        ~~~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:47:49: error: no member named 'TransformType' in namespace 'o2'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                                            ~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:47:73: error: no member named 'TransformType' in namespace 'o2'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                                                                    ~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:48:49: error: no member named 'TransformType' in namespace 'o2'
                                            o2::TransformType::L2G)); // request cached transforms
                                            ~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:54:15: error: use of undeclared identifier 'Hit'
  std::vector<Hit>* hitArray = nullptr;
              ^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:54:21: error: C++ requires a type specifier for all declarations
  std::vector<Hit>* hitArray = nullptr;
                    ^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:61:15: error: use of undeclared identifier 'Cluster'
  std::vector<Cluster>* clusArray = nullptr;
              ^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:61:25: error: C++ requires a type specifier for all declarations
  std::vector<Cluster>* clusArray = nullptr;
                        ^
Error in <TInterpreter::AutoParse>: Error parsing payload code for class o2::Rotation2D with content:
--

```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/867/378e7b83097db8555e3a8f289e609afc6785f259/fullLog.txt).

alibuild commented:
Error while checking build/o2/macos for 378e7b83097db8555e3a8f289e609afc6785f259:
```
sw/BUILD/O2-latest/log
      getSinkSpec() in test_DataAllocator.cxx.o
      o2::framework::DataProcessorSpec::~DataProcessorSpec() in test_DataAllocator.cxx.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [bin/test_Framework_test_DataAllocator] Error 1
make[1]: *** [Framework/Core/CMakeFiles/test_Framework_test_DataAllocator.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/867/378e7b83097db8555e3a8f289e609afc6785f259/fullLog.txt).

alibuild commented:
Error while checking build/o2/macos for 7b27062986aa25669cbbf93591970af6491c8d50:
```
sw/BUILD/O2-latest/log
      getSinkSpec() in test_DataAllocator.cxx.o
      o2::framework::DataProcessorSpec::~DataProcessorSpec() in test_DataAllocator.cxx.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [bin/test_Framework_test_DataAllocator] Error 1
make[1]: *** [Framework/Core/CMakeFiles/test_Framework_test_DataAllocator.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/867/7b27062986aa25669cbbf93591970af6491c8d50/fullLog.txt).

alibuild commented:
Error while checking build/O2/o2-dev-fairroot for 7b27062986aa25669cbbf93591970af6491c8d50:
```
sw/BUILD/O2-latest/log
#undef  _BACKWARD_BACKWARD_WARNING_H

In file included from input_line_10:1:
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:24:19: error: no member named 'Cluster' in namespace 'o2::ITSMFT'
using o2::ITSMFT::Cluster;
      ~~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/7e2770c80a55ac267e8e5007c180a91fef4be384/O2/lib/libITSMFTSimulation.rootmap declaring '
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:31:21: error: no member named 'Hit' in namespace 'o2::ITSMFT'
  using o2::ITSMFT::Hit;
        ~~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/7e2770c80a55ac267e8e5007c180a91fef4be384/O2/lib/libMFTBase.rootmap declaring '
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:46:24: error: no member named 'GeometryTGeo' in namespace 'o2::MFT'
  auto gman = o2::MFT::GeometryTGeo::Instance();
              ~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:47:36: error: no member named 'bit2Mask' in namespace 'o2::utils'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                        ~~~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:47:49: error: no member named 'TransformType' in namespace 'o2'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                                            ~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:47:73: error: no member named 'TransformType' in namespace 'o2'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                                                                    ~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:48:49: error: no member named 'TransformType' in namespace 'o2'
                                            o2::TransformType::L2G)); // request cached transforms
                                            ~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:54:15: error: use of undeclared identifier 'Hit'
  std::vector<Hit>* hitArray = nullptr;
              ^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:54:21: error: C++ requires a type specifier for all declarations
  std::vector<Hit>* hitArray = nullptr;
                    ^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:61:15: error: use of undeclared identifier 'Cluster'
  std::vector<Cluster>* clusArray = nullptr;
              ^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:61:25: error: C++ requires a type specifier for all declarations
  std::vector<Cluster>* clusArray = nullptr;
                        ^
Error in <TInterpreter::AutoParse>: Error parsing payload code for class o2::Rotation2D with content:
--

```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/867/7b27062986aa25669cbbf93591970af6491c8d50/fullLog.txt).

alibuild commented:
Error while checking build/o2/macos for 89aafb1285d54bb99848698d4f5fb9714e3620e5:
```
sw/BUILD/O2-latest/log
      getSinkSpec() in test_DataAllocator.cxx.o
      o2::framework::DataProcessorSpec::~DataProcessorSpec() in test_DataAllocator.cxx.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [bin/test_Framework_test_DataAllocator] Error 1
make[1]: *** [Framework/Core/CMakeFiles/test_Framework_test_DataAllocator.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/867/89aafb1285d54bb99848698d4f5fb9714e3620e5/fullLog.txt).

alibuild commented:
Error while checking build/o2/macos for 53a7f7de92acd3e69e1c91d2404ac08acfeda45f:
```
sw/BUILD/O2-latest/log
      getSinkSpec() in test_DataAllocator.cxx.o
      o2::framework::DataProcessorSpec::~DataProcessorSpec() in test_DataAllocator.cxx.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [bin/test_Framework_test_DataAllocator] Error 1
make[1]: *** [Framework/Core/CMakeFiles/test_Framework_test_DataAllocator.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/867/53a7f7de92acd3e69e1c91d2404ac08acfeda45f/fullLog.txt).

alibuild commented:
Error while checking build/O2/o2-dev-fairroot for 0c618c5041b720f6b1ba9c2a579890f8755f7662:
```
sw/BUILD/O2-latest/log
#undef  _BACKWARD_BACKWARD_WARNING_H

In file included from input_line_10:1:
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/ITS/macros/test/CheckClusters.C:26:21: error: no member named 'Cluster' in namespace 'o2::ITSMFT'
  using o2::ITSMFT::Cluster;
        ~~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/7e2770c80a55ac267e8e5007c180a91fef4be384/O2/lib/libITSMFTSimulation.rootmap declaring '
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/ITS/macros/test/CheckClusters.C:27:21: error: no member named 'Hit' in namespace 'o2::ITSMFT'
  using o2::ITSMFT::Hit;
        ~~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/7e2770c80a55ac267e8e5007c180a91fef4be384/O2/lib/libITSBase.rootmap declaring '
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/ITS/macros/test/CheckClusters.C:39:24: error: no member named 'GeometryTGeo' in namespace 'o2::ITS'
  auto gman = o2::ITS::GeometryTGeo::Instance();
              ~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/ITS/macros/test/CheckClusters.C:40:36: error: no member named 'bit2Mask' in namespace 'o2::utils'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2GRot,
                        ~~~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/ITS/macros/test/CheckClusters.C:40:49: error: no member named 'TransformType' in namespace 'o2'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2GRot,
                                            ~~~~^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/ITS/macros/test/CheckClusters.C:40:73: error: no member named 'TransformType' in namespace 'o2'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2GRot,
                                                                    ~~~~^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/ITS/macros/test/CheckClusters.C:41:49: error: no member named 'TransformType' in namespace 'o2'
                                            o2::TransformType::L2G)); // request cached transforms
                                            ~~~~^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/ITS/macros/test/CheckClusters.C:47:15: error: use of undeclared identifier 'Hit'
  std::vector<Hit>* hitArray = nullptr;
              ^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/ITS/macros/test/CheckClusters.C:47:21: error: C++ requires a type specifier for all declarations
  std::vector<Hit>* hitArray = nullptr;
                    ^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/ITS/macros/test/CheckClusters.C:54:15: error: use of undeclared identifier 'Cluster'
  std::vector<Cluster>* clusArr = nullptr;
              ^
/mnt/mesos/sandbox/sandbox/O2/Detectors/ITSMFT/ITS/macros/test/CheckClusters.C:54:25: error: C++ requires a type specifier for all declarations
  std::vector<Cluster>* clusArr = nullptr;
                        ^
Error in <TInterpreter::AutoParse>: Error parsing payload code for class o2::Rotation2D with content:
--

```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/867/0c618c5041b720f6b1ba9c2a579890f8755f7662/fullLog.txt).

alibuild commented:
Error while checking build/o2/macos for 0c618c5041b720f6b1ba9c2a579890f8755f7662:
```
sw/BUILD/O2-latest/log
      getSinkSpec() in test_DataAllocator.cxx.o
      o2::framework::DataProcessorSpec::~DataProcessorSpec() in test_DataAllocator.cxx.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [bin/test_Framework_test_DataAllocator] Error 1
make[1]: *** [Framework/Core/CMakeFiles/test_Framework_test_DataAllocator.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/867/0c618c5041b720f6b1ba9c2a579890f8755f7662/fullLog.txt).

matthiasrichter commented:
I'm leaving out one commit for the moment until I understand the compilation issue on mac. I have included all reasonable changes by `clang-format`. I suggest we merge the PR.

alibuild commented:
Error while checking build/O2/o2-dev-fairroot for ca8ff30a99736218a93f7ee88d4ef4ba7adb2923:
```
sw/BUILD/O2-latest/log
#undef  _BACKWARD_BACKWARD_WARNING_H

In file included from input_line_10:1:
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:24:19: error: no member named 'Cluster' in namespace 'o2::ITSMFT'
using o2::ITSMFT::Cluster;
      ~~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/7e2770c80a55ac267e8e5007c180a91fef4be384/O2/lib/libITSMFTSimulation.rootmap declaring '
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:31:21: error: no member named 'Hit' in namespace 'o2::ITSMFT'
  using o2::ITSMFT::Hit;
        ~~~~~~~~~~~~^
Warning in <TInterpreter::LoadLibraryMap>: Problems in /mnt/mesos/sandbox/sandbox/sw/BUILD/7e2770c80a55ac267e8e5007c180a91fef4be384/O2/lib/libMFTBase.rootmap declaring '
--

#undef  _BACKWARD_BACKWARD_WARNING_H

/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:46:24: error: no member named 'GeometryTGeo' in namespace 'o2::MFT'
  auto gman = o2::MFT::GeometryTGeo::Instance();
              ~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:47:36: error: no member named 'bit2Mask' in namespace 'o2::utils'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                        ~~~~~~~~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:47:49: error: no member named 'TransformType' in namespace 'o2'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                                            ~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:47:73: error: no member named 'TransformType' in namespace 'o2'
  gman->fillMatrixCache(o2::utils::bit2Mask(o2::TransformType::T2L, o2::TransformType::T2G,
                                                                    ~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:48:49: error: no member named 'TransformType' in namespace 'o2'
                                            o2::TransformType::L2G)); // request cached transforms
                                            ~~~~^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:54:15: error: use of undeclared identifier 'Hit'
  std::vector<Hit>* hitArray = nullptr;
              ^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:54:21: error: C++ requires a type specifier for all declarations
  std::vector<Hit>* hitArray = nullptr;
                    ^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:61:15: error: use of undeclared identifier 'Cluster'
  std::vector<Cluster>* clusArray = nullptr;
              ^
/mnt/mesos/sandbox/sandbox/O2/macro/CheckClusters_mft.C:61:25: error: C++ requires a type specifier for all declarations
  std::vector<Cluster>* clusArray = nullptr;
                        ^
Error in <TInterpreter::AutoParse>: Error parsing payload code for class o2::Rotation2D with content:
--

```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/867/ca8ff30a99736218a93f7ee88d4ef4ba7adb2923/fullLog.txt).
